server/utils/deferred.ts

```diff
@@ -0,0 +1,40 @@
+type Callback = () => unknown;
+type DeferredTask = () => Promise<void>;
+
+const createDeferredState = () => {
+	let deferredCount = 0;
+	let deferredCallbacks: Callback[] = [];
+
+	const incrementCount = () => {
+		deferredCount++;
+	};
+
+	const decrementCount = () => {
+		deferredCount--;
+		if (deferredCount === 0) {
+			deferredCallbacks.forEach((cb) => cb());
+			deferredCallbacks = [];
+		}
+	};
+
+	const onDeferredTasksComplete = (callback: Callback) => {
+		if (deferredCount === 0) {
+			callback();
+		} else {
+			deferredCallbacks.push(callback);
+		}
+	};
+
+	const finishDeferredTasks = (): Promise<void> => {
+		return new Promise((resolve) => onDeferredTasksComplete(resolve));
+	};
+
+	const defer = (task: DeferredTask) => {
+		incrementCount();
+		task().then(decrementCount);
```

Comment:
I guess `defer` will be in the call stack so it should be clear that an uncaught error was the result of a deferred task. But we could add a catch here to be more deliberate about it? idk

Reply:
I'm into the idea of doing something here, though I don't know what would be most appropriate. Catch and print the stack trace?

Reply:
Yeah I'm not exactly sure, maybe something like:

```ts
.catch(err => {
  console.error("An error occurred in a deferred task:")
  console.error(err)
})
```

Would at least give us a string to grep for if we needed it?